Uniswap LP Bridge: Updated testing, and fixed linting errors.

ndebrouwer opened the pull request:
I fixed the linting errors as requested, and have also updated the foundry tests to work with the new test base. I want to continue to improve the tests also to check other cases for the bridge to be safe, but as I understand Jan had some feedback, I wanted him to see the new updated version so he can review it so I can start working on those suggestions ASAP. 

My personal ideas for things I want to add/check:
-add some foundry tests to check that the bridge correctly reverts in the case of improper aztecasset inputs and outputs.
-perhaps delete the mintbyswap convert method, as now that Jan has created the Uniswap Swap bridge, the user can instead swap using that and then call my bridge to LP.
-Make sure the tests ,which currently only test the dai/weth pool, generalize well to other pools
-Find out why calling convert through the rolluprocessor fails. At the moment, I am impersonating the rollupprocessor using vm.prank and then calling my bridge directly. For some reason, calling through rollupprocessor triggers a revert in the Uniswap V3 Pool. I think the delegatecall the rollupprocessor performs is the culprit for some reason, but I'm not completely sure. 



# Checklist:

- [ ] I have reviewed my diff in github, line by line.
- [ ] Every change is related to the PR description.
- [ ] There are no unexpected formatting changes, or superfluous debug logs.
- [ ] The branch has been rebased against the head of its merge target.
- [ ] I'm happy for the PR to be merged at the reviewers next convenience.

benesjan commented on src/bridges/uniswapv3/SyncUniswapV3Bridge.sol:
```diff
@@ -0,0 +1,449 @@
+     * @return _auxData the packed _auxData
+     */
+
+    function packData(
```
Please remove the blank line between function definition and natspec everywhere.

benesjan commented on src/bridges/uniswapv3/AsyncUniswapV3Bridge.sol:
```diff
@@ -0,0 +1,330 @@
+            bool isAsync
+        )
+    {
+        if (!(msg.sender == address(ROLLUP_PROCESSOR) || msg.sender == address(this))) revert InvalidCaller();
```
Why is it allowed for the msg.sender to be address(this) here? This just seems like a pointless introduction of potential reentrancy vulnerability

benesjan commented on src/test/bridges/uniswapv3/WorkingAsync.t.sol:
```diff
@@ -0,0 +1,394 @@
+        //_setTokenBalance(address(DAI), address(ROLLUP_PROCESSOR), depositAmount, 2);
+        uint64 data = _packData(uint160(80), int24(20), int24(20));
+
+        if (true) {
```
Why is there if (true) here?

benesjan commented on src/test/bridges/uniswapv3/WorkingAsync.t.sol:
```diff
@@ -0,0 +1,394 @@
+            );
+            vm.stopPrank();
+        }
+        if (false) {
```
Why is there if (false) here?

benesjan commented on src/test/bridges/uniswapv3/WorkingAsync.t.sol:
```diff
@@ -0,0 +1,394 @@
+        return true;
+
+        vm.assume(_deposit >= 1000000000000000);
+        vm.assume(_deposit <= 60048070258392067512707354988);
```
please replace these 2 assume calls with 1 bound(...) call from forge-std.

ndebrouwer commented on src/aztec/interfaces/IRollupProcessor.sol:
```diff
@@ -111,4 +113,21 @@ interface IRollupProcessor {
 
     function getSupportedBridges() external view returns (address[] memory, uint256[] memory);
+
+    function convert(
```
I did this temporarily because I was thinking that convert was called on rollupProcessor still and my fxn call wasn't working because it wasn't defined in interface. Realized after the change was on purpose and used sendDefiRollup instead. I have fixed this.

ndebrouwer commented on src/bridges/uniswapv3/AsyncUniswapV3Bridge.sol:
```diff
@@ -0,0 +1,330 @@
+        address _nonfungiblePositionManager,
+        address _factory,
+        address _wEth,
+        address _quoter
```
Agreed, hardcoded them.

ndebrouwer commented on src/bridges/uniswapv3/interfaces/IERC165.sol:
```diff
@@ -0,0 +1,25 @@
+ *
+ * For an implementation, see {ERC165}.
+ */
+interface IERC165 {
```
done

ndebrouwer commented:
Hey Jan, I addressed every comment here and made some improvements. I replied to your comments but I didn't want to blow your up notifications or be repetitive; but I did indeed check every one and fixed them. Most of them were idiosyncrasies left over from testing that I should've gotten rid of and now have.

The circle test failed on the foundry tests but I just checked and it wasn't my tests that made it fail so mine are good to go as far as that is concerned. 

I look forward to any further comments on your part. 

Additionally, I have two primary questions which I would like to speak to you about.

1) At the moment my tests are passing but I am calling my bridge directly. Whenever I tried to get the tests to work by doing it through the rollupProcessor, it reverts. I suspect this is due to the delegatecall the rollupProcessor makes because I encoutered this issue in the past. If we can figure this out so I can verify calling through the rollupProcessor will have the same expected behavior, that is great. I expect the behavior should be exactly as calling through the bridge directly but I just want to be sure. 

2) Now that the Uniswap bridge for swapping is up, I question whether it is still relevant for my bridge to have a convert call that allows the user to swap. This is interaction 3 of the UniLPBridge, in tests it is testMintBySwap(). I did this due to the fact that sometimes a user may only have one asset and need to swap it to get the two assets they need to provide liquidity, but since there was no swap bridge, I needed a solution so they wouldn't have to exit the L2 network, swap, re-enter, etc. So maybe it is not so relevant now due to the swap bridge, maybe it is. Unsure. I would like your thoughts

benesjan commented:
@ndebrouwer We've updated sendDefiRollup(...) function and now it returns the same values as bridge's convert(...) function. Should make it much easier to write the E2E tests.

ndebrouwer commented:
> @ndebrouwer We've updated `sendDefiRollup(...)` function and now it returns the same values as bridge's `convert(...)` function. Should make it much easier to write the E2E tests.

Hey Jan, thanks for letting me know. I used your suggestion and increased the gas limit for the bridge btw and it worked, so that problem is solved, thank you! 

Also: in response to me asking about swapping on my LP bridge, you said "Hello, this should be handled by the bridge". Just to be clear, you mean the official Aztec Connect bridge for swaping, or my LP bridge? 

Finally, I have run into the following problem that I would like to discuss with you. You mentioned the refunds and I thought about it and I agreed initially it would be better to have the bridge swap the refunds so that maximal liquidity possible is deposited. So I have been working on that in addition to finishing touches on my bridge.

First I must ask though: which helper contracts are you referring to that can deposit the funds for the user and swap them until maximal liquidity is achieved, because I had trouble finding any

Because I couldn't find any, I wrote some code to try achieve maximal liquidity for the user but I ran into a problem.
I used this paper : 	https://atiselsts.github.io/pdfs/uniswap-v3-liquidity-math.pdf as a reference for the math to find optimal liquidity. x(x√P ·√p_b)·(√p_b −√P )=y(√P −√p_a)

The problem is that this equation tells you given some x, what y you need to achieve optimal liquidity. However, the user cannot just conjure any y they wish to have, they have an initial supply of x and y which is finite. Therefore we basically have the constraint that 
y = (S-x)*r where r is the ratio of token1 (Y) / token0 (X). In other words our y is constrained by how much x we can swap to get y. 
       The problem with this however is that it assumes that token0 and token1 are just linearly convertible. This assumption is safe at reasonable levels of token0 because slippage is low but at high levels it begins to break down because big swaps change price/have slippage obviously. Still, if we treat the assumption as safe, by substituting y = (S-x)*r, 
with A = (x√P ·√p_b)·(√p_b −√P )/(√P −√p_a)
we can find a solution. x = Sr/ (A+r). 

	So at low reasonable token0 qty we can indeed find a better allocation of x and y if it exists. But at high  amounts we can't without doing alot of checking because slippage breaks the assumptions necessary to find the solution. I implemented checking before actually swapping to ensure that the optimal x found by our solution still works even after the price shifts. We can do this by using QuoterV2 to simulate both the expected output after we swap to optimal quantities and the price after swapping. then we can check if the liquidity would be greater even with these altered parameters than compared to before. And sometimes it doesn't improve liquidity to swap, it can reduce it. intuitively if we lack a token and need to swap to get more of it, it gets more expensive after we swap for it, but since if a token is more expensive, we need more of it provide liquidity. need token -> swap for it -> token gets more expensive -> need more token to provide same level of liquidity as before. It becomes a 'dog chasing its own tail' sort of deal. There is probably some optimal compromise, but to simulate that is gas intensive. And it adds alot of complexity and gas costs even in the easy case where the token quantities are low. 

	Therefore, I feel that while it is still useful to provide the user with an option to find the optimal quantities for max liquidity and then swap to that automatically for them, it would be better to provide that as a helper function that the user can call offchain,
so that there are 0 extra added gas costs. once the user has found optimal quantities, they can call the swap bridge and swap. I think this compromise makes the most sense. Do you agree or disagree ?

benesjan commented:
Hello @ndebrouwer,

> Also: in response to me asking about swapping on my LP bridge, you said "Hello, this should be handled by the bridge". Just to be clear, you mean the official Aztec Connect bridge for swaping, or my LP bridge?

Yes, by that I mean that the bridge would ideally always deposit the input tokens in full.

> First I must ask though: which helper contracts are you referring to that can deposit the funds for the user and swap them until maximal liquidity is achieved, because I had trouble finding any

I could not find the contracts after a quick search as well but I remember that there was some project on twitter which did this with 1 input token. This allowed for 1 sided liquidity provision. I think having only 1 token on the input would also be ok. I asked on the Uniswap Grantees slack and I will let you know once I get a response.

BTW can you please make CI pass? It seems there are some broken imports. Thank you

benesjan commented:
Grant finished
